### PR TITLE
Show error message after crash

### LINKED ISSUE #4108: Prevent crashes when there are errors talking to SQLite database
We run into a few different crashes when trying to read or write to/from the database on device:

https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/5974a495deaaafc213c6857985557684?time=last-ninety-days&versions=v2021.2.2%20(4256)&sessionEventKey=610BA6A002AE00014A23B626FE362FE0_1571311994101282197

https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/b88555ad2701253e0177019be276acff?time=last-ninety-days&versions=v2021.2.2%20(4256)&sessionEventKey=610BA5FE01140001472DB626FE362FE0_1571311629015343902

This a mixture of the database being full or not accessible. There isn't much we can do to recover from these but we could give the user some instructions:

* `SQLiteFullException` - "Device storage is full. Try deleting some things and then try again"
* `SQLiteReadOnlyDatabaseException` - "Collect's database is corrupt. Try submitting forms and then reinstalling."

### LINKED ISSUE #5145: Problems with creating files on device storage
#### Problem description
We have been receiving reports about problems with accessing app directory which we use to save files. In v2022.2.2 we added more logs and here are the results:

- [open failed: EACCES (Permission denied) | No such file or directory | Operation not permitted | open failed: EROFS (Read-only file system) | Permission denied](https://console.firebase.google.com/u/1/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/4ff8a6e3b84fc921539f2592eba88c84?time=last-seven-days&versions=v2022.2.2%20(4437)&types=crash&sessionEventKey=6283DE3D00B000010624EAA39C8B8CE8_1677207395531930656)
- [External files dir is null!](https://console.firebase.google.com/u/1/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/535e9fed953082e894f6263be94462b0?time=last-seven-days&versions=v2022.2.2%20(4437)&types=crash&sessionEventKey=628752C902DE00011E60B926C195CE47_1678179966787820882)
- [External files dir does not exist: /storage/emulated/0/Android/data/org.odk.collect.android/files](https://console.firebase.google.com/u/1/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/a8eebdc52ac37e3c247a8fc13cc8d3d0?time=last-seven-days&versions=v2022.2.2%20(4437)&types=crash&sessionEventKey=62811BE302EB00012DE6191AA7D0D4D5_1676430775496630780)

@kkrawczyk123 was able to reproduce one of those issues (see https://forum.getodk.org/t/can-t-create-directory-and-application-crashes/36583/15?u=grzesiek2010)
I took that device from here so I can take a look at it to see if there is anything we can do.

### PR DESCRIPTION
**Pull request by seadowg**
Closes #5145 
Closes #4108 

The above issues aren't strictly "fixed", but we're changing how we deal with crashes and so I think it's worth closing adn then readdressing them if we feel like this crash UI isn't enough later.

The app will now display a "crash screen" on startup with details (the message) of an exception in two scenarios:

1. The app crashed last time it was run
2. The app was unable to start properly (currently only due to an inability to access external storage)

To make this easier to verify, I've added a "Crash app" button in a new "Developer tools" section of Experimental settings. There's not a particularly good way to verify **2** without messing with code which I think the reviewer should do.

#### What has been done to verify that this works as intended?

New tests.

#### Why is this the best possible solution? Were any other approaches considered?

I also looked at showing a dialog/screen when the app crashes (rather than on next launch), but this appeared to be not possible without doing some shady things (like creating an alarm in Android to relaunch the app).

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

I think the only things to check here are to try out the new "Crash app" button and check that the flow when relaunching (or after a restart) all works without anything weird happening. 

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: androidshared/src/main/res/layout/crash_layout.xml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    <androidx.constraintlayout.widget.Guideline
+        android:id="@+id/guideline_bottom"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        app:layout_constraintGuide_end="@dimen/margin_standard" />
+
+    <com.google.android.material.textview.MaterialTextView
+        android:id="@+id/title"
+        style="?textAppearanceTitleLarge"
```
Comment (seadowg) [inline androidshared/src/main/res/layout/crash_layout.xml]: I'm using Material 3 styles in this file. We've already updated `STATE.md` to say we should use them wherever they fit in, but just wanted to highlight.

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import androidx.activity.OnBackPressedCallback
+import org.odk.collect.crash_handler.CrashHandler
+import org.odk.collect.strings.localization.LocalizedActivity
+
+class CrashHandlerActivity : LocalizedActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        val crashHandler = CrashHandler.getInstance(this)!!
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt:13]: Not usually a fan of `!!`, but we really shouldn't be getting to this Activity if the `CrashHandler` instance is `null`.

File: collect_app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -98,7 +98,11 @@ the specific language governing permissions and limitations under the License.
         </provider>
 
         <activity
-            android:name=".activities.FirstLaunchActivity"
+            android:name=".activities.FirstLaunchActivity">
+        </activity>
+
+        <activity
+            android:name=".activities.MainMenuActivity"
```
Comment (seadowg) [inline collect_app/src/main/AndroidManifest.xml:105]: Switching to the `MainMenuActivity` is a pretty big change. I needed to make that happen to account for the Android's ["restart on crash" behaviour](https://www.seadowg.com/2022/10/24/crash-test.html).

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java [hunk 4]
```diff
@@ -129,15 +121,21 @@ protected void attachBaseContext(Context base) {
...
         singleton = this;
-        setupDagger();
-        DaggerUtils.getComponent(this).inject(this);
 
-        applicationInitializer.initialize();
-        fixGoogleBug154855417();
-        setupStrictMode();
+        CrashHandler crashHandler = CrashHandler.install(this);
+        boolean shouldLaunch = crashHandler.checkConditions(() -> {
+            ExternalFilesUtils.testExternalFilesAccess(this);
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: We might want to put other checks in here to prevent other crashes from happening down the road.

File: strings/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -1159,4 +1159,23 @@
     -->
     <string name="open_settings">Open Settings</string>
 
+    <!--
+    ##############################################
+    # Crash handling
+    ##############################################
+    -->
+
+    <string name="crash_last_run">There was a crash last time the app was running!</string>
```
Comment (seadowg) [inline strings/src/main/res/values/strings.xml:1175]: I'm not particularly happy with the copywriting here. @lognaturel can you suggest some alternatives?

**Comment by seadowg** [comment]
@lognaturel we probably also want to revise the message for the storage crash. The user will currently see the exception messages that come out of `ExternalFilesUtils.testExternalFilesAccess`.

File: crash-handler/build.gradle [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import dependencies.Dependencies
```
Comment (grzesiek2010) [inline crash-handler/build.gradle]: As we agreed recently - we should start using kotlin in new gradle files.

File: .circleci/test_modules.txt [hunk 7]
```diff
@@ -13,3 +13,4 @@ permissions:testDebug
 settings:testDebug
 maps:testDebug
 errors:testDebug
+crash-handler:testDebug
```
Comment (grzesiek2010) [inline .circleci/test_modules.txt:16]: So far we haven't used dashes in module names bu there is nothing wrong about it I guess?

File: androidshared/src/main/res/layout/crash_layout.xml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        android:layout_marginHorizontal="@dimen/margin_extra_large"
+        android:gravity="center"
+        android:textAlignment="center"
+        android:layout_marginTop="@dimen/margin_standard"
+        app:layout_constraintEnd_toEndOf="@id/guideline_end"
+        app:layout_constraintStart_toStartOf="@id/guideline_start"
+        app:layout_constraintTop_toBottomOf="@id/title"
+        tools:text="It exploded!" />
+
+    <Button
```
Comment (grzesiek2010) [inline androidshared/src/main/res/layout/crash_layout.xml]: You should add `contentDescription` to this button.

File: androidshared/src/main/res/layout/crash_layout.xml [hunk 9]
```diff
@@ -0,0 +1,84 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <androidx.constraintlayout.widget.Guideline
```
Comment (grzesiek2010) [inline androidshared/src/main/res/layout/crash_layout.xml]: I think you don't need all those guidelines. You can just add `android:padding="@dimen/margin_standard"` to the main container and constraint every element that used those guidelines to parent.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/pages/MainMenuPage.java [hunk 10]
```diff
@@ -27,7 +27,8 @@ public class MainMenuPage extends Page<MainMenuPage> {
     @Override
     public MainMenuPage assertOnPage() {
         return WaitFor.waitFor(() -> {
-            onView(withText(StringContains.containsString(getTranslatedString(R.string.collect_app_name)))).perform(scrollTo()).check(matches(isDisplayed()));
+            onView(withText(R.string.enter_data)).check(matches(isDisplayed()));
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/pages/MainMenuPage.java:30]: Do we need both checks here (R.string.enter_data string and app name)?

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt [hunk 11]
```diff
@@ -56,11 +57,19 @@ class CollectTestRule @JvmOverloads constructor(
...
     ): Instrumentation.ActivityResult {
         val scenario = launchForResult<Activity>(intent)
         destination.assertOnPage()
         actions.accept(destination)
         return scenario.result
     }
+
+    private fun getLaunchIntent(): Intent {
+        val context = ApplicationProvider.getApplicationContext<Context>()
+        val intent = context.packageManager.getLaunchIntentForPackage("org.odk.collect.android")
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: You should be able to use `BuildConfig.APPLICATION_ID` instead of that `org.odk.collect.android` right? And generally it could look like:
```
    private fun getLaunchIntent(): Intent {
        return ApplicationProvider
            .getApplicationContext<Context>()
            .packageManager
            .getLaunchIntentForPackage(BuildConfig.APPLICATION_ID)!!
            .apply {
                this.addCategory(Intent.CATEGORY_LAUNCHER)
            }
    }
```

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt [hunk 12]
```diff
@@ -0,0 +1,28 @@
+package org.odk.collect.android.activities
+
+import android.os.Bundle
+import androidx.activity.OnBackPressedCallback
+import org.odk.collect.crash_handler.CrashHandler
+import org.odk.collect.strings.localization.LocalizedActivity
+
+class CrashHandlerActivity : LocalizedActivity() {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt:8]: Why it's not in `crash-handler` module?

File: collect_app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/AndroidManifest.xml:105]: If it's necessary to make a change like this one I think it would be better to have a separate staring activity and rename `FirstLaunchActivity` to something like `FirstProjectConfiguratorActivity`.

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java [hunk 13]
```diff
@@ -129,15 +121,21 @@ protected void attachBaseContext(Context base) {
...
-
         singleton = this;
-        setupDagger();
-        DaggerUtils.getComponent(this).inject(this);
 
-        applicationInitializer.initialize();
-        fixGoogleBug154855417();
-        setupStrictMode();
+        CrashHandler crashHandler = CrashHandler.install(this);
+        boolean shouldLaunch = crashHandler.checkConditions(() -> {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Could be:
`        boolean shouldLaunch = crashHandler.checkConditions(() -> ExternalFilesUtils.testExternalFilesAccess(this));
`

File: crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+        private const val KEY_CRASH = "crash"
+
+        @JvmStatic
+        fun install(context: Context): CrashHandler {
+            val crashHandler = CrashHandler()
+            wrapUncaughExceptionHandler(crashHandler, context)
+
+            return crashHandler.also {
+                context.getState().set("crash_handler", it)
```
Comment (grzesiek2010) [inline crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt:91]: This "crash_handler" should be a const.

File: crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }
+
+        @JvmStatic
+        fun getInstance(context: Context): CrashHandler? {
+            return context.getState().get<CrashHandler>("crash_handler")
+        }
+
+        private fun wrapUncaughExceptionHandler(
```
Comment (grzesiek2010) [inline crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt]: You missed `t`

File: collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java [hunk 16]
```diff
@@ -77,16 +80,30 @@ public class MainMenuActivity extends CollectAbstractActivity {
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
-        new ThemeUtils(this).setDarkModeForCurrentProject();
-
+        SplashScreen.installSplashScreen(this);
         super.onCreate(savedInstanceState);
+
+        CrashHandler crashHandler = CrashHandler.getInstance(this);
+        if (crashHandler != null && crashHandler.hasCrashed(this)) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java:98]: Is it even possible that's it null?

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java (hunk 13 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: and it doesn't even need to return any value. It could have another `Runnable` as a parameter to run setup if everything is fine. Like:
```
        crashHandler.checkConditions(() -> {
            ExternalFilesUtils.testExternalFilesAccess(this);
        }, () -> {
            setupDagger();
            DaggerUtils.getComponent(this).inject(this);

            applicationComponent.applicationInitializer().initialize();
            fixGoogleBug154855417();
            setupStrictMode();
        });
```
Probably we would need to set the name of that method then too.

File: crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+import android.content.Context
+import org.odk.collect.androidshared.data.getState
+import kotlin.system.exitProcess
+
+class CrashHandler(private val processKiller: Runnable = Runnable { exitProcess(0) }) {
+
+    private var conditionFailure: String? = null
+
+    fun registerCrash(context: Context, crash: Throwable) {
+        getPreferences(context).edit().putString(KEY_CRASH, crash.message ?: "").apply()
```
Comment (grzesiek2010) [inline crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt]: Shouldn't we keep this in meta prefs?

File: crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+    private fun getPreferences(context: Context) =
+        context.getSharedPreferences("crash_handler", Context.MODE_PRIVATE)
+
+    companion object {
+
+        private const val KEY_CRASH = "crash"
+
+        @JvmStatic
+        fun install(context: Context): CrashHandler {
+            val crashHandler = CrashHandler()
```
Comment (grzesiek2010) [inline crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt]: I think it would make more sense to pass `crashHandler` to app state immediately here:
```
           val crashHandler = CrashHandler().also {
                context.getState().set("crash_handler", it)
            }
```

File: crash-handler/src/main/java/org/odk/collect/crash_handler/CrashView.kt [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+
+    private var onErrorDismissed: Runnable? = null
+
+    init {
+        inflate(context, R.layout.crash_layout, this)
+        findViewById<View>(R.id.ok_button).setOnClickListener { dismiss() }
+    }
+
+    fun setCrash(title: String, message: String?, onErrorDismissed: Runnable? = null) {
+        findViewById<TextView>(R.id.title).text = title
```
Comment (grzesiek2010) [inline crash-handler/src/main/java/org/odk/collect/crash_handler/CrashView.kt]: Can we use binding instead of all those `findViewById`?

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+        val crashView = crashHandler.getCrashView(this) {
+            ActivityUtils.startActivityAndCloseAllOthers(this, MainMenuActivity::class.java)
+        }!!
+
+        setContentView(crashView)
+
+        onBackPressedDispatcher.addCallback(
+            object : OnBackPressedCallback(true) {
+                override fun handleOnBackPressed() {
+                    crashView.dismiss()
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt]: Don't wee need a test to verify it's called?

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

**Comment by grzesiek2010** [comment]
My final thought is that we should turn the `OK` button into `Send feedback`.
Screenshot_1667577303
That button would move a user to a new view something like:
image

or to make it simpler we could immediately combine those two vies into one displaying that alert icon and some explanation on top and the rest below. 
But it could be a part of v2023.1

File: crash-handler/build.gradle (hunk 6 above)
Comment (seadowg) [reply crash-handler/build.gradle]: Ah yes! We had that conversation after I'd created this and then I forgot to go back for it. Will fix.

File: .circleci/test_modules.txt (hunk 7 above)
Comment (seadowg) [reply .circleci/test_modules.txt:16]: Yeah, I had thought you couldn't do it, but it seems totally fine! Definitely nicer. We can look at renaming other modules at some point.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/pages/MainMenuPage.java (hunk 10 above)
Comment (seadowg) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/pages/MainMenuPage.java:30]: Yes because both pieces of text appear on other screens. They only appear together on the main menu (as far as I'm aware).

File: androidshared/src/main/res/layout/crash_layout.xml (hunk 8 above)
Comment (seadowg) [reply androidshared/src/main/res/layout/crash_layout.xml]: I'm pretty sure `Button` will generate a content description based on the `text` that screen readers will then use. As far as I'm aware, we only need `contentDescription` on views that don't have text (or custom views).

File: collect_app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (seadowg) [reply collect_app/src/main/AndroidManifest.xml:105]: An Activity like that would need to either `finish()` when launching the next Activity or have `noHistory` which is what causes the problem when the app restarts (for a crash or because of a memory reclaim): the Activity is not in the backstack that's restored.

File: androidshared/src/main/res/layout/crash_layout.xml (hunk 8 above)
Comment (grzesiek2010) [reply androidshared/src/main/res/layout/crash_layout.xml]: Right, it extends `TextView` so i works out of the box.

File: .circleci/test_modules.txt (hunk 7 above)
Comment (grzesiek2010) [reply .circleci/test_modules.txt:16]: What about using the underscore sign? We have it in `collect_app` module. It would be better to use one convention across the app.

Comment (seadowg) [reply .circleci/test_modules.txt:16]: I really prefer `-`, but I don't want to rename `collect_app` 😂. It would be nice for that module to just be `app` (as is the convention). I could just rename this module `crash` so we can punt on that right now?

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt (hunk 12 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt:8]: It references `MainMenuActivity`. I wanted to create some isolated crash handling tools that we could easily move to a Fragment or Compose based UI later so ended up with the Activity living in `collect_app` and handling the navigation.

File: crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt (hunk 17 above)
Comment (seadowg) [reply crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt]: Yeah I'd initially though about it working that way (or creating a new "temp" settings). The problem is that we can't pass/inject a `SettingsProvider` from `collect_app` the Dagger graph (the `AppDependencyComponent`) might not have been created when crashes occur or are displayed. Given that, I decided to have `CrashHandler` just use raw Android APIs and give it its own `SharedPreferences` bucket. Does that make sense?

File: collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java (hunk 16 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java:98]: Yes. See https://github.com/getodk/collect/pull/5324/commits/3a29c6adf632e947b863b20ea60bc82596855019. We could definitely make switching `CrashHandler` "on and off" more explicit for tests, but it is one of the advantages of using `AppState` that we get fresh runs every time.

File: .circleci/test_modules.txt (hunk 7 above)
Comment (grzesiek2010) [reply .circleci/test_modules.txt:16]: You can leave `crash-handler` but please file a separate issue to rename other modules.

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt (hunk 20 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt]: Good catch. Will add that.

File: crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt (hunk 17 above)
Comment (grzesiek2010) [reply crash-handler/src/main/java/org/odk/collect/crash_handler/CrashHandler.kt]: Yes but maybe we should add a comment I'm sure I will forget the reason soon and will be thinking about the same next time I came across this code.

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt (hunk 12 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt:8]: So maybe we could start `MainMenuActivity` without referencing it directly? For example by adding a custom category/intent name? You know I'm talking about marking the `MainMenuActivity` as the main activity and then start without using its class name.

**Comment by grzesiek2010** [comment]
@seadowg you have some conflicts to fix too.

File: .circleci/test_modules.txt (hunk 7 above)
Comment (seadowg) [reply .circleci/test_modules.txt:16]: Done: https://github.com/getodk/collect/issues/5337

**Comment by seadowg** [comment]
> My final thought is that we should turn the OK button into Send feedback.

@grzesiek2010 totally agree that we want to get some kind of call to action in there. Wed also briefly discussed linking to the forum as an example. As you say, I think that'll end up being an enhancement we drop in v2023.1.

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java (hunk 13 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: I went back and forth on versions like this, but couldn't decide what was best. Will go with the version you suggested.

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt (hunk 12 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt:8]: Totally see what you're saying, and it does make sense to be able to launch the "launch"/"start"/"home" Activity without an explicit reference. I started playing around with that, and then changed my mind: I still think the `CrashHandlerActivity` should live outside the `crash-handler` module and be a client of it rather than part of it. The reasoning for me here comes back to my earlier comment: 

> I wanted to create some isolated crash handling tools that we could easily move to a Fragment or Compose based UI later

Basically, it feels to me that (although these kinds of things are always guesses) that we're probably likely to end up changing navigation to the crash UI separately to changing the design itself. With that assumption, it makes sense to have `crash-handler` deal with recording and the actual UI elements, but let the client (here `CrashHandlerActiivty`) dictate the navigation as part of an actual app module (`collect_app`). Does that make sense?

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java [hunk 21]
```diff
@@ -136,15 +129,21 @@ protected void attachBaseContext(Context base) {
...
-
         singleton = this;
-        setupDagger();
-        DaggerUtils.getComponent(this).inject(this);
 
-        applicationInitializer.initialize();
-        fixGoogleBug154855417();
-        setupStrictMode();
+        CrashHandler.install(this).launchApp(
+                () -> {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: It's just one line so you do not need `{}`

File: collect_app/src/test/java/org/odk/collect/android/activities/CrashHandlerActivityTest.kt [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+    private val context = ApplicationProvider.getApplicationContext<Context>()
+
+    @get:Rule
+    val launcherRule = ActivityScenarioLauncherRule()
+
+    @Test
+    fun `pressing back dismisses crash view`() {
+        val crashHandler = CrashHandler.getInstance(context)
+        crashHandler!!.createMockViews = true
+        crashHandler.registerCrash(context, RuntimeException("BAM"))
```
Comment (grzesiek2010) [inline collect_app/src/test/java/org/odk/collect/android/activities/CrashHandlerActivityTest.kt:30]: This would look better I think: 
```
CrashHandler.getInstance(context)?.also {
            it.createMockViews = true
            it.registerCrash(context, RuntimeException("BAM"))
        }
```

File: collect_app/src/test/java/org/odk/collect/android/activities/CrashHandlerActivityTest.kt [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+        launcherRule.launch(CrashHandlerActivity::class.java).onActivity {
+            it.onBackPressedDispatcher.onBackPressed()
+
+            val crashView = it.getContextView<MockCrashView>()
+            assertThat(crashView.wasDismissed, equalTo(true))
```
Comment (grzesiek2010) [inline collect_app/src/test/java/org/odk/collect/android/activities/CrashHandlerActivityTest.kt:38]: It would be good to assert it was false before.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt (hunk 12 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt:8]: I'm not entirely convinced but I also do not have strong feelings to implement it in the way I described so ok.

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java [hunk 24]
```diff
@@ -93,17 +91,12 @@ public class Collect extends Application implements
...
-
-    @Inject
-    SettingsProvider settingsProvider;
-
     private AudioRecorderDependencyComponent audioRecorderDependencyComponent;
     private ProjectsDependencyComponent projectsDependencyComponent;
     private GeoDependencyComponent geoDependencyComponent;
     private OsmDroidDependencyComponent osmDroidDependencyComponent;
     private EntitiesDependencyComponent entitiesDependencyComponent;
+    public CrashHandler crashHandler;
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: This is unused.

File: collect_app/src/test/java/org/odk/collect/android/activities/CrashHandlerActivityTest.kt (hunk 22 above)
Comment (seadowg) [reply collect_app/src/test/java/org/odk/collect/android/activities/CrashHandlerActivityTest.kt:30]: Good idea. I'll keep a `!!` in instead of a `?`, as the test wouldn't make sense if `CrashHandler.getInstance(context)` returned `null`.

File: collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt:13]: I'm still thinking about it... can't we make sure that `CrashHandler` always exist? We always should have that object available right?

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/activities/CrashHandlerActivity.kt:13]: Currently, no. It wouldn't be available before calling `install` and it really shouldn't be available after calling `uninstall` (I'll make that change now). As I said in https://github.com/getodk/collect/pull/5324#discussion_r1017838264, we currently use the fact that it can be nulled out by clearing app state as well.

**Review by grzesiek2010 — APPROVED** [review]

**Comment by dbemke** [comment]
On Android 10- Redmi 9T the "crash app” options causes continuous crashes of the app when the dark mode is applied. First, there is a dialog about crash details - simulated crash. After other crashes following dialogs are with null pointer exception.
seconddialog

**Steps to reproduce:**
1. In Collect settings choose dark theme.
2. Go to Experimental→ Developer tools → Crash app
3. In the dialog "ODK Colect has stopped" tap "cancel".
4. Go back to the Collect app.
5. When you see the dialog about the crash tap "ok”.

After those steps the user is moved to step 3 – so the user gets to a loop of steps 3,4,5.

**Other information:**
On Android 13 steps 1-5 reproduce but don’t cause the ongoing crashes.

**Comment by srujner** [comment]
On Android 13 there is some issue with Theme styling: 

Steps to Reproduce: 
1. Have the Light mode set in Device
2. Have the Dark mode set in the Collect App
3. Go to the Settings -> Experimental -> Dev tools -> Crash app
4. Click on OK button

Other information:
- Changing Theme Style in Settings should be enough to reproduce it
- I was not able to reproduce it on Master 

Screenshot_20221114-122927

https://user-images.githubusercontent.com/85225519/201686114-2875262b-3ac7-4e90-a2b8-83575013f2c9.mp4

**Comment by seadowg** [comment]
Looks like both issues are probably related and are due to some real weirdness in how the "theme is different from app" feature. Just so I don't forget: what's happening is that the dark theme is being set when we open `MainMenuActivity` after view the crash, that restarts **every Activity in the backstack** which unfortunately includes `CrashHandlerActivity` while `MainMenuActivity#onCreate` runs 😣.

**Comment by seadowg** [comment]
@dbemke @srujner the crash UI shouldn't cause crashes when the theme is different from the system one now. You'll still see that the crash screen will use the system UI regardless of what the app has set up. This is because that screen needs to be able to be shown in cases we haven't been able to load the user's settings. Furthermore, the screen isn't really "project specific" so it isn't clear project theme we should use.

The specific problem @srujner found with the button's loading incorrect themes looks like it's probably https://github.com/getodk/collect/issues/4836. Were you seeing that when just changing the theme in the app? Nothing to do with using the new crash UI? And was it happening every time?

**Comment by seadowg** [comment]
@dbemke @srujner I think I should have fixed the Android 13 dark/light theme problem. @grzesiek2010 would be good to take a look and see if [what I did](https://github.com/getodk/collect/pull/5324/commits/1b2678b47f79acb212807729d781bcaf71b9e9e1) makes sense to you. I thought calling `SplashScreen.installSplashScreen` was meant to be a compat method that'd be safe on any OS version, but it seemed to cause problems for the dark/light theme switching. It might be that they assume you have a specific Activity for the Splash Screen (like we did), but that would mean you'd get a blank screen on restarts (as the Activity would have to be removed from history). Maybe that's OK though? Either way, my changes seemed to have fixed the issue and not broken the Splash Screen on older and newer versions.

**Comment by dbemke** [comment]
**Tested with Success!**
Verified on Android 10, 13 (physical device and the last commit on the emulator)

**Verified Cases:**
- issue on Android 13 with wrong theme applied and issue on Android 10 with continuous crashes are no longer reproducing
- simulated crash
- real crashes of the app (with don’t keep activities enabled and disabled) and checking the "crash screen" on startup with details (the message) of an exception
- simulating a crash when a list of forms is loading
- in device settings "Always show crash dialog” enabled and disabled
- dark/light mode
- rotating, minimizing and locking the screen on the crash screen
- after a crash choosing open the app again or close the app - Android 8.1
- crashing the app from a shortcut and from an opened form
